src/h5web/visualizations/heatmap/utils.ts

```diff
@@ -23,3 +28,67 @@ export function computeTextureData(matrix: number[][]): Uint8Array | undefined {
 
   return Uint8Array.from(colors.flat());
 }
+
+export function usePanZoom(camera: OrthographicCamera): ReactThreeFiber.Events {
+  const startOffsetPosition = useRef<Vector3>(); // `useRef` to avoid re-renders
+
+  const onPointerDown = useCallback(
+    (evt: PointerEvent) => {
+      evt.stopPropagation();
+
+      const { currentTarget, pointerId } = evt as React.PointerEvent;
+      currentTarget.setPointerCapture(pointerId);
+
+      const projectedPoint = camera.worldToLocal(evt.unprojectedPoint.clone());
+      startOffsetPosition.current = camera.position.clone().add(projectedPoint);
+    },
+    [camera]
+  );
+
+  const onPointerUp = useCallback((evt: PointerEvent) => {
+    evt.stopPropagation();
+
+    const { currentTarget, pointerId } = evt as React.PointerEvent;
+    currentTarget.releasePointerCapture(pointerId);
+
+    startOffsetPosition.current = undefined;
+  }, []);
+
+  const onPointerMove = useCallback(
+    (evt: PointerEvent) => {
+      if (!startOffsetPosition.current) {
+        return;
+      }
+
+      evt.stopPropagation();
+
+      const projectedPoint = camera.worldToLocal(evt.unprojectedPoint.clone());
+      const { x: pointX, y: pointY } = projectedPoint;
```

Comment:
The two above lines cannot be put in one ?

Reply:
It can, but it then wraps onto three lines... 😂 Also, I thought the variable added clarity.